### ISSUE TITLE
Asan leaks in Vulkan

### ISSUE THREAD

**Issue by GMNGeoffrey** [google/iree#5716]
We have leaks detected by asan (well lsan really) in our vulkan tests. For now I'm disabling vulkan from the asan bot I'm bringing up. I'm not sure if this is something we can fix, at least for swiftshader, or whether we just need to figure out the right sanitizer suppressions.

Logs: https://gist.github.com/13f0f1d8bf176ec8c0d04145fc92e021

**Comment by benvanik** [comment]
So long as IREE_ASSERT_OK is tied to gtest it's fine to wrap in an iree::Status - if we wanted to switch to a C unit testing framework then we'd want to make it safe.

We could try making the macros something like:
```c
#define IREE_ASSERT_OK(rexpr) ASSERT_THAT(iree::Status(rexpr), ..)
```
or
```c
#define IREE_ASSERT_OK(rexpr) \
  { \
    iree::Status __status = {(rexpr)}; \
    ASSERT_THAT(__status...) \
  }
```
though we may want to do some `&&`  trickery to ensure they aren't used later; this would be bad:
```c
iree_status_t status = ...;
IREE_ASSERT_OK(status);
do_something(status);  // status not valid here
```

All for making it harder to leak them in tests in the normal path --- leaks in failure cases are fine (as barely any test code is expected to gracefully fail).

**Comment by ScottTodd** [comment]
Yeah, the status leaks came up when I broke dlopen by passing the wrong flags - I finally saw accurate asan stacks, but for status code and not Vulkan code :P

Any dlopen()/dlclose() tweaks may need to reach into the Vulkan loader as well: https://github.com/KhronosGroup/Vulkan-Loader/blob/8198bebc7fe31c3da54b1dfacbb92e8697646701/loader/vk_loader_platform.h#L238-L257. I'm also trying some LD_PRELOAD tricks but no luck so far.

**Comment by ScottTodd** [comment]
Editing the Vulkan Loader code linked above helped me get useful reports from ASAN, which all pointed to leaks throughout `libnvidia-glcore.so` on my machine. When I switch to using SwiftShader, I see no leaks in IREE's Vulkan tests.

Possible next steps:
* Debug inside our Docker images, with dlopen/dlclose changes to IREE's code and the Vulkan Loader
* Update our SwiftShader build to a more recent commit - maybe there were leaks that were fixed
* Investigate ASAN suppressions (these _might_ also need the dlopen/dlclose changes?)
